### DISCUSSION TITLE
Connect to an Azure Service Bus queue and topic.

### DISCUSSION THREAD

**Discussion by chankly** [rebus-org/Rebus#1161] [Q&A]
Hello !!
Surely I am misunderstanding some concept of the bookstore.

I'm trying to make a simple example where my system connects to a queue and a topic. I manage to connect to the queue, and receive the message (The queue is even created automatically and another queue called "error", its normal?). But the topic is neither created and nor works.

In my IoC I have this implemented:

> builder.Services.AddRebus((cfg, sp) =>
> {
>     return cfg.Routing(cfg =>
>                 {
>                     var routing = cfg.TypeBased();
>                     routing.Map<UserCreatedEvent>("myqueue2");
>                     routing.Map<UserCreatedTopicEvent>("mytopic2");
>                 })
>               .Transport(t => t.UseAzureServiceBus("XXXXXXXXXXXXXXXXXXXXXXXXX", "myqueue2"));
> });
> builder.Services.AddRebusHandler< UserCreatedEventHandler >();
> builder.Services.AddRebusHandler< UserCreatedTopicEventHandler >();

In my test controller I have this implemented:

> [HttpGet()]
>         public async Task<IActionResult> Get([FromServices] IBus _bus)
>         {
>             await _bus.Send(new UserCreatedEvent("queue2"));
>             await _bus.Publish(new UserCreatedTopicEvent("topic2"));
>             return Ok();
>         }

When I execute the Action of the controller, a topic is created with an ugly name but never I received the message: webapplication2~webapplication2.messaging.usercreatedtopicevent

Can any kind soul give me some light?

**Comment by mookid8000** [comment]
> The queue is even created automatically and another queue called "error", its normal?

Yes it's normal. 🙂 

> But the topic is neither created and nor works.

Here's how to work with topic-based routing (a.k.a. "pub/sub") with Rebus. 

(1) First you start all the "participants", i.e. a Rebus instance for each system that should be involved in the communication. Each of them must have its own input queue.

(2) Then you ensure that someone subscribes to something:

```csharp
await bus.Subscribe<SomethingInterestingHappened>();
```

This will create a topic called "SomethingInterestingHappened" (not exactly that, though, because by default there will be more info in it).

When that is done, the topic will have a binding to the subscriber's input queue. This is materialized by Rebus in Azure Service Bus concepts by creating a "subscription" beneath the topic, which will be configured to forward events from the topic to the subscriber's input queue.

(3) Last, start publishing events.

Any participant can now

```csharp
await bus.Publish(new SomethingInterestingHappened(...));
```

and a copy will be received by each participant that has subscribed.

That's how pub/sub messaging works with Rebus.

Btw. you'll only need endpoint mappings when you want to `await bus.Send(new SomeCommand(..))`, e.g. by calling `.Map<SomeCommand>("some-queue-name")` , which will cause the command to be sent to the queue "some-queue-name".

I hope that makes it clearer 🙂 You can also read about it here: https://github.com/rebus-org/Rebus/wiki/Routing

**Reply by chankly** [reply]
Sorry for not having answered. Thank you very much for your answer. The answer has been helpful to me. :)
